contracts/op-finality-gadget/src/exec/finality.rs

```diff
@@ -328,13 +307,84 @@ fn check_fp_exist(deps: Deps, fp_pubkey_hex: &str) -> Result<(), ContractError>
     }
 }
 
+/// `slash_finality_provider` slashes a finality provider with the given evidence including setting
+/// its voting power to zero, extracting its BTC SK, and emitting an event
+fn slash_finality_provider(
+    env: &Env,
+    fp_btc_pk_hex: &str,
+    evidence: &Evidence,
+) -> Result<(WasmMsg, Event), ContractError> {
+    let pk = eots::PublicKey::from_hex(fp_btc_pk_hex)?;
+    let btc_sk = pk
+        .extract_secret_key(
+            &evidence.pub_rand,
+            &evidence.canonical_app_hash,
+            &evidence.canonical_finality_sig,
+            &evidence.fork_app_hash,
+            &evidence.fork_finality_sig,
+        )
+        .map_err(|err| ContractError::SecretKeyExtractionError(err.to_string()))?;
+
+    // Emit slashing event.
+    // Raises slashing event to babylon over IBC.
+    let msg = ExecuteMsg::Slashing {
+        evidence: evidence.clone(),
+    };
+    let wasm_msg: WasmMsg = WasmMsg::Execute {
+        contract_addr: env.contract.address.to_string(),
+        msg: to_json_binary(&msg)?,
+        funds: vec![],
+    };
+
+    let ev = Event::new("slashed_finality_provider")
+        .add_attribute("module", "finality")
+        .add_attribute("finality_provider", fp_btc_pk_hex)
+        .add_attribute("block_height", evidence.block_height.to_string())
+        .add_attribute(
+            "canonical_app_hash",
+            hex::encode(&evidence.canonical_app_hash),
+        )
+        .add_attribute(
+            "canonical_finality_sig",
+            hex::encode(&evidence.canonical_finality_sig),
+        )
+        .add_attribute("fork_app_hash", hex::encode(&evidence.fork_app_hash))
+        .add_attribute(
+            "fork_finality_sig",
+            hex::encode(&evidence.fork_finality_sig),
+        )
+        .add_attribute("secret_key", hex::encode(btc_sk.to_bytes()));
+    Ok((wasm_msg, ev))
+}
+
+pub(crate) fn handle_slashing(evidence: &Evidence) -> Result<Response<BabylonMsg>, ContractError> {
+    let mut res = Response::new();
+    // Send msg to Babylon
+
+    let msg = BabylonMsg::EquivocationEvidence {
+        evidence: Some(evidence.clone()),
+    };
```

Comment:
Probably the syntax needs to be fixed as per https://github.com/babylonlabs-io/babylon/pull/327